src/nodebuilder.cpp

```diff
@@ -69,11 +70,29 @@ void NodeBuilder::OnMapStart(const Mark& mark, const std::string& tag,
   node.set_tag(tag);
   node.set_style(style);
   m_mapDepth++;
+  m_mergeDicts.emplace_back();
+}
+
+void MergeMapCollection(detail::node& map_to, detail::node& map_from,
+                        detail::shared_memory_holder& pMemory) {
+  const detail::node& const_map_to = map_to;
+  for (auto j = map_from.begin(); j != map_from.end(); j++) {
+    detail::node* s = const_map_to.get(*j->first, pMemory);
```

Comment:
This compares the `shared_ptr` of the keys, so it will not work correctly, this can be visible by iterating over keys, here is a fix for this case - https://github.com/jbeder/yaml-cpp/pull/1279

It is based on your PR, with this fix and a test on top.